daft/ai/utils.py

```diff
@@ -0,0 +1,21 @@
+from __future__ import annotations
+
+import torch
+
+
+def get_device() -> torch.device:
+    """Get the best available PyTorch device for computation.
+
+    This function automatically selects the optimal device in order of preference:
+    1. CUDA GPU (if available) - for NVIDIA GPUs with CUDA support
+    2. MPS (Metal Performance Shaders) - for Apple Silicon Macs
+    3. CPU - as fallback when no GPU acceleration is available
+    """
+    device = (
+        torch.device("cuda")
+        if torch.cuda.is_available()
+        else torch.device("mps")
+        if hasattr(torch.backends, "mps") and torch.backends.mps.is_available()
+        else torch.device("cpu")
+    )
+    return device
```

Comment:
```suggestion
from __future__ import annotations


def get_device() -> torch.device:
    """Get the best available PyTorch device for computation.

    This function automatically selects the optimal device in order of preference:
    1. CUDA GPU (if available) - for NVIDIA GPUs with CUDA support
    2. MPS (Metal Performance Shaders) - for Apple Silicon Macs
    3. CPU - as fallback when no GPU acceleration is available
    """
    import torch
    
    device = (
        torch.device("cuda")
        if torch.cuda.is_available()
        else torch.device("mps")
        if hasattr(torch.backends, "mps") and torch.backends.mps.is_available()
        else torch.device("cpu")
    )
    return device
```

lazy-import torch so utils can contain non-torch methods. It's valuable to have this here since it applies to both transformers and sentence_transformers. It would also make sense to unify those two providers to just 'transformers' and then this helper would only be in transformers, but I want to get this work in and we could clean up later.

Reply:
yeah smashing transformers together makes sense to me